src/traits.rs

```diff
@@ -848,11 +878,491 @@ pub trait PrimInt
+   /// significant 1 bit set.
+   ///
+   /// # Examples
+   ///
+   /// ```
+   /// use num::traits::PrimInt;
+   ///
+   /// let n = 0b0101_0000u8;
+   /// let s = 0b0001_1111u8;  
+   ///
+   /// assert_eq!(n.mask_trailing_zeros_and_least_significant_one(), s);
+   /// ```
+   fn mask_trailing_zeros_and_least_significant_one(self) -> Self {
+       (self - cast(1).unwrap()) ^ self
+   }
+
+   /// Returns a mask with all of the trailing 1's of `self` set and the least
+   /// significant 0 bit set.
+   ///
+   /// # Examples
+   ///
+   /// ```
+   /// use num::traits::PrimInt;
+   ///
+   /// let n = 0b0101_1111u8;
+   /// let s = 0b0011_1111u8;  
+   ///
+   /// assert_eq!(n.mask_trailing_ones_and_least_significant_zero(), s);
+   /// ```
+   fn mask_trailing_ones_and_least_significant_zero(self) -> Self {
+       self ^ (self + cast(1).unwrap())
+   }
+
+   fn general_reverse_bits(self, subword_bits: u32, group_subwords: u32) -> Self;
+
+   /// Reverses the bits of `self`
+   ///
+   /// # Examples
+   ///
+   /// ```
+   /// use num::traits::PrimInt;
+   ///
+   /// let n = 0b1011_0010u8;
+   /// let s = 0b0100_1101u8;  
+   /// assert_eq!(n.reverse_bits(), s);
+   ///
+   /// let n1 = 0b1011_0010_1010_1001u16;
+   /// let s1 = 0b1001_0101_0100_1101u16;
+   /// assert_eq!(n1.reverse_bits(), s1);
+   /// ```
+   fn reverse_bits(self) -> Self {
+       self.general_reverse_bits(1, 1)
+   }
+
+   /// Reverses the pairs of bits of `self`
+   ///
+   /// # Examples
+   ///
+   /// ```
+   /// use num::traits::PrimInt;
+   ///
+   /// let n = 0b1011_0010u8;
+   /// let s = 0b1000_1110u8;  
+   /// assert_eq!(n.reverse_bit_pairs(), s);
+   ///
+   /// let n1 = 0b1011_0010_1010_1001u16;
+   /// let s1 = 0b0110_1010_1000_1110u16;
+   /// assert_eq!(n1.reverse_bit_pairs(), s1);
+   /// ```
+   fn reverse_bit_pairs(self) -> Self {
+       self.general_reverse_bits(2, 1)
+   }
+
+   /// Reverses the nibbles of `self`
+   ///
+   /// # Examples
+   ///
+   /// ```
+   /// use num::traits::PrimInt;
+   ///
+   /// let n = 0b1011_0010u8;
+   /// let s = 0b0010_1011u8;  
+   /// assert_eq!(n.reverse_bit_nibbles(), s);
+   ///
+   /// let n1 = 0b1011_0010_1010_1001u16;
+   /// let s1 = 0b1001_1010_0010_1011u16;
+   /// assert_eq!(n1.reverse_bit_nibbles(), s1);
+   /// ```
+   fn reverse_bit_nibbles(self) -> Self {
+       self.general_reverse_bits(4, 1)
+   }
+
+   fn reverse_byte_groups(self, bytes_per_block: u32, blocks_per_group: u32) -> Self {
+       self.general_reverse_bits(8 * bytes_per_block, blocks_per_group)
+   }
+
+   /// Reverses the bytes of `self`
+   ///
+   /// # Examples
+   ///
+   /// ```
+   /// use num::traits::PrimInt;
+   ///
+   /// let n = 0b1011_0010u8;
+   /// let s = 0b1011_0010u8;  
+   /// assert_eq!(n.reverse_bytes(), s);
+   ///
+   /// let n1 = 0b1011_0010_1010_1001u16;
+   /// let s1 = 0b1010_1001_1011_0010u16;
+   /// assert_eq!(n1.reverse_bytes(), s1);
+   /// ```
+   fn reverse_bytes(self) -> Self {
```

Comment:
Yes it is, I've added a note, an assert, and forwarded it to `swap_bytes` since it should use the llvm `swapbs` intrinsics which is faster. 

I don't know if I should remove this function or leave it though. For consistency with the other bit reversal algorithms it makes sense for it to exist and users might wonder why not, but those looking for the CPU / LLVM intrinsic will look for `swap_bytes`. However the other reversal algorithms are not called `swap_bits` or similar. 

Need to give more thought to this.